### PR TITLE
[SPARK-43965][PYTHON][CONNECT] Support Python UDTF in Spark Connect

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?

This PR supports creating and registering Python UDTFs in Spark Connect.

### Why are the changes needed?

To make Python UDTF work with Spark Connect.

### Does this PR introduce _any_ user-facing change?

Yes. After this PR, users can use Python UDTFs in Spark Connect.

### How was this patch tested?

New unit tests

### CODE REVIEW

File: python/pyspark/sql/connect/udtf.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    def __call__(self, *cols: "ColumnOrName") -> "DataFrame":
+        from pyspark.sql.connect.dataframe import DataFrame
+        from pyspark.sql.connect.session import _active_spark_session
+
+        if _active_spark_session is None:
+            # TODO: can we create a Spark session here?
+            raise PySparkRuntimeError(
+                "An active SparkSession is required for "
+                "executing a Python user-defined table function."
+            )
```
Comment (allisonwang-db) [inline python/pyspark/sql/connect/udtf.py:167]: @HyukjinKwon @zhengruifeng Given that we don't have `SparkSession._getActiveSessionOrCreate()` in Spark Connect, what's the best approach to creating a Spark Connect session when the active Spark session is None? Or, should we not recommend this and instead throw an exception when the Spark Connect session is None?

**Comment by allisonwang-db** [comment]
cc @HyukjinKwon @zhengruifeng @xinrong-meng

File: python/pyspark/sql/connect/udtf.py (hunk 1 above)
Comment (zhengruifeng) [reply python/pyspark/sql/connect/udtf.py:167]: currently, we use `pyspark.sql.connect.session._active_spark_session` in ml and pandas

https://github.com/apache/spark/blob/10751dc285c5c639e3343a8abc26857407522822/python/pyspark/ml/util.py#L756-L758

https://github.com/apache/spark/blob/b97ce8b9a99c570fc57dec967e7e9db3d115c1db/python/pyspark/pandas/utils.py#L484-L486

Comment (zhengruifeng) [reply python/pyspark/sql/connect/udtf.py:167]: oh, we don't have `getActiveSessionOrCreate` now

**Review by zhengruifeng — APPROVED** [review]
Looks pretty good, but I want leave it to @HyukjinKwon @xinrong-meng to sign off

File: connector/connect/common/src/main/protobuf/spark/connect/relations.proto [hunk 2]
```diff
@@ -941,6 +942,36 @@ message ApplyInPandasWithState {
...
+
+  // (Optional) Whether the user-defined table function is deterministic.
+  bool deterministic = 2;
+
+  // (Optional) Function input arguments. Empty arguments are allowed.
+  repeated Expression arguments = 3;
+
+  // (Required) Type of the user-defined table function.
+  oneof function {
+    PythonUDTF python_udtf = 4;
```
Comment (xinrong-meng) [inline connector/connect/common/src/main/protobuf/spark/connect/relations.proto:957]: Out of curiosity will there be UDTFs in other languages?

File: python/pyspark/sql/connect/functions.py [hunk 3]
```diff
@@ -3891,6 +3894,21 @@ def udf(
 udf.__doc__ = pysparkfuncs.udf.__doc__
 
 
+def udtf(
+    cls: Optional[Type] = None,
+    *,
+    returnType: Union[StructType, str],
+    useArrow: Optional[bool] = None,
+) -> Union["UserDefinedTableFunction", functools.partial]:
```
Comment (xinrong-meng) [inline python/pyspark/sql/connect/functions.py:3902]: We typically do not use `functools.partial` as a type hint in Python. Shall we use `Callable` following the `udf` return type hints?

File: connector/connect/common/src/main/protobuf/spark/connect/relations.proto (hunk 2 above)
Comment (allisonwang-db) [reply connector/connect/common/src/main/protobuf/spark/connect/relations.proto:957]: Currently only Python but more UDTFs might be supported in the future.

File: python/pyspark/sql/connect/plan.py [hunk 4]
```diff
@@ -2173,6 +2175,107 @@ def plan(self, session: "SparkConnectClient") -> proto.Relation:
...
+                method="ddl_parse", ddl_string=self._return_type.data_type_string
+            ).parsed
+        else:
+            parsed = self._return_type
+
+        if not isinstance(parsed, StructType):
+            raise PySparkTypeError(
+                error_class="INVALID_UDTF_RETURN_TYPE",
+                message_parameters={"name": self._name, "return_type": f"{parsed}"},
+            )
```
Comment (ueshin) [inline python/pyspark/sql/connect/plan.py:2208]: As this check is done in the server side, we don't need to check this in client side, then we can reduce one round-trip to parse the schema. `UnparsedDataType` will be handled in `SparkConnectPlanner.transformDataType`.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by zhengruifeng** [comment]
@allisonwang-db the test failures are related

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**